client/state/post-types/reducer.js

```diff
@@ -0,0 +1,74 @@
+/**
+ * External dependencies
+ */
+import { combineReducers } from 'redux';
+import indexBy from 'lodash/collection/indexBy';
+
+/**
+ * Internal dependencies
+ */
+import { isValidStateWithSchema } from 'state/utils';
+import * as schema from './schema';
+import {
+	SERIALIZE,
+	DESERIALIZE,
+	POST_TYPES_RECEIVE,
+	POST_TYPES_REQUEST,
+	POST_TYPES_REQUEST_SUCCESS,
+	POST_TYPES_REQUEST_FAILURE
+} from 'state/action-types';
+
+/**
+ * Returns the updated requests state after an action has been dispatched. The
+ * state maps site ID keys to whether a request for post types is in progress.
+ *
+ * @param  {Object} state  Current state
+ * @param  {Object} action Action payload
+ * @return {Object}        Updated state
+ */
+export function requesting( state = {}, action ) {
+	switch ( action.type ) {
+		case POST_TYPES_REQUEST:
+		case POST_TYPES_REQUEST_SUCCESS:
+		case POST_TYPES_REQUEST_FAILURE:
+			return Object.assign( {}, state, {
+				[ action.siteId ]: POST_TYPES_REQUEST === action.type
+			} );
+
+		case SERIALIZE:
+		case DESERIALIZE:
+			return {};
+	}
+
+	return state;
+}
+
+/**
+ * Returns the updated items state after an action has been dispatched. The
+ * state maps site ID keys to an object of post type name to post type for all
+ * supported post types on the site.
+ *
+ * @param  {Object} state  Current state
+ * @param  {Object} action Action payload
+ * @return {Object}        Updated state
+ */
+export function items( state = {}, action ) {
+	switch ( action.type ) {
+		case POST_TYPES_RECEIVE:
+			return Object.assign( {}, state, {
+				[ action.siteId ]: indexBy( action.types, 'name' )
+			} );
+
+		case DESERIALIZE:
+			if ( ! isValidStateWithSchema( state, schema.items ) ) {
+				return {};
+			}
```

Comment:
>Maybe add a comment that this is supposed to fall through. We might run into an error if another case is added below DESERIALIZE.

Ah, I hadn't meant for the total fall-through, should have incorporated a break. The "fall through" should be that it only returns the initial state if invalid, otherwise `return state;` at the bottom of the reducer takes effect. Verges a bit far on the "too clever" side, so I think I'll just rebase to omit d121e2da4222690074a645ec0fbc033df5b69e5c from the history.